url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/network/datacenter/Switch.java

```diff
@@ -0,0 +1,386 @@
+
+	public ArrayList<NetworkPacket> pktlist;
+
+	public List<Vm> BagofTaskVm = new ArrayList<Vm>();
+
+	public double switching_delay;
+
+	public Map<Integer, NetworkVm> Vmlist;
+
+	public Switch(String name, int level, NetworkDatacenter dc) {
+		super(name);
+		this.level = level;
+		this.dc = dc;
+	}
+
+	@Override
+	public void startEntity() {
+		Log.printLine(getName() + " is starting...");
+		schedule(getId(), 0, CloudSimTags.RESOURCE_CHARACTERISTICS_REQUEST);
+	}
+
+	@Override
+	public void processEvent(SimEvent ev) {
+		// Log.printLine(CloudSim.clock()+"[Broker]: event received:"+ev.getTag());
+		switch (ev.getTag()) {
+		// Resource characteristics request
+			case CloudSimTags.Network_Event_UP:
+				// process the packet from down switch or host
+				processpacket_up(ev);
+				break;
+			case CloudSimTags.Network_Event_DOWN:
+				// process the packet from uplink
+				processpacket_down(ev);
+				break;
+			case CloudSimTags.Network_Event_send:
+				processpacketforward(ev);
+				break;
+
+			case CloudSimTags.Network_Event_Host:
+				processhostpacket(ev);
+				break;
+			// Resource characteristics answer
+			case CloudSimTags.RESOURCE_Register:
+				registerHost(ev);
+				break;
+			// other unknown tags are processed by this method
+			default:
+				processOtherEvent(ev);
+				break;
+		}
+	}
+
+	protected void processhostpacket(SimEvent ev) {
+		// Send packet to host
+		NetworkPacket hspkt = (NetworkPacket) ev.getData();
+		NetworkHost hs = hostlist.get(hspkt.recieverhostid);
+		hs.packetrecieved.add(hspkt);
+	}
+
+	protected void processpacket_down(SimEvent ev) {
+		// packet coming from up level router.
+		// has to send downward
+		// check which switch to forward to
+		// add packet in the switch list
+		// add packet in the host list
+		// int src=ev.getSource();
+		NetworkPacket hspkt = (NetworkPacket) ev.getData();
+		int recvVMid = hspkt.pkt.reciever;
+		CloudSim.cancelAll(getId(), new PredicateType(CloudSimTags.Network_Event_send));
+		schedule(getId(), latency, CloudSimTags.Network_Event_send);
+		if (level == NetworkConstants.EDGE_LEVEL) {
+			// packet is to be recieved by host
+			int hostid = dc.VmtoHostlist.get(recvVMid);
+			hspkt.recieverhostid = hostid;
+			List<NetworkPacket> pktlist = packetTohost.get(hostid);
+			if (pktlist == null) {
+				pktlist = new ArrayList<NetworkPacket>();
+				packetTohost.put(hostid, pktlist);
+			}
+			pktlist.add(hspkt);
+			return;
+		}
+		if (level == NetworkConstants.Agg_LEVEL) {
+			// packet is coming from root so need to be sent to edgelevel swich
+			// find the id for edgelevel switch
+			int switchid = dc.VmToSwitchid.get(recvVMid);
+			List<NetworkPacket> pktlist = downlinkswitchpktlist.get(switchid);
+			if (pktlist == null) {
+				pktlist = new ArrayList<NetworkPacket>();
+				downlinkswitchpktlist.put(switchid, pktlist);
+			}
+			pktlist.add(hspkt);
+			return;
+		}
+
+	}
+
+	protected void processpacket_up(SimEvent ev) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 174-204, lines 205-233, lines 234-258 into separate methods.